src/pmp.c

```diff
@@ -88,56 +75,51 @@ void metal_pmp_init(struct metal_pmp *pmp) {
     };
 
     for (unsigned int i = 0; i < _pmp_regions(); i++) {
-        metal_pmp_set_region(pmp, i, init_config, 0);
+        metal_pmp_set_region(i, init_config, 0);
     }
 
     /* Detect the region granularity by writing all 1s to pmpaddr0 while
      * pmpcfg0 = 0. */
-    if (metal_pmp_set_address(pmp, 0, -1) != 0) {
+    if (metal_pmp_set_address(0, -1) != 0) {
         /* Failed to detect granularity */
         return;
     }
 
     /* Calculate the granularity based on the value that pmpaddr0 takes on */
-    pmp->_granularity[metal_cpu_get_current_hartid()] =
-        _get_detected_granularity(metal_pmp_get_address(pmp, 0));
+    _granularity[metal_cpu_get_current_hartid()] =
+        _get_detected_granularity(metal_pmp_get_address(0));
 
     /* Clear pmpaddr0 */
-    metal_pmp_set_address(pmp, 0, 0);
+    metal_pmp_set_address(0, 0);
 }
 
-int metal_pmp_set_region(struct metal_pmp *pmp, unsigned int region,
-                         struct metal_pmp_config config, size_t address) {
+int metal_pmp_set_region(unsigned int region, struct metal_pmp_config config,
+                         size_t address) {
     struct metal_pmp_config old_config;
     size_t old_address;
     size_t cfgmask;
     size_t pmpcfg;
     int rc = 0;
 
-    if (!pmp) {
-        /* Device handle cannot be NULL */
-        return 1;
-    }
-
     if (region > _pmp_regions()) {
         /* Region outside of supported range */
         return 2;
     }
 
     if (config.A == METAL_PMP_NA4 &&
-        pmp->_granularity[metal_cpu_get_current_hartid()] > 4) {
+        _granularity[metal_cpu_get_current_hartid()] > 4) {
```

Comment:
This should be in the DTS; we shouldn't be doing run-time detection of this value. Sounds like a bug should be filed against the DTS generation?